File: vigenere.js
function k_combinations(set, k) {
    /**
     * Copyright 2012 Akseli Palén.
     * Created 2012-07-15.
     * Licensed under the MIT license.
    **/

    var i, j, combs, head, tailcombs;

    if (k > set.length || k <= 0) {
        return [];
    }

    if (k == set.length) {
        return [set];
    }

    if (k == 1) {
        combs = [];
        for (i = 0; i < set.length; i++) {
            combs.push([set[i]]);
        }
        return combs;
    }

    combs = [];
    for (i = 0; i < set.length - k + 1; i++) {

        head = set.slice(i, i + 1);

        tailcombs = k_combinations(set.slice(i + 1), k - 1);

        for (j = 0; j < tailcombs.length; j++) {
            combs.push(head.concat(tailcombs[j]));
        }
    }
    return combs;
}

function* get_factors(x) {
    for (let i = 2; i < Math.floor(x / 2) + 1; i++) {
        if (x % i === 0) {
            yield i;
        }
    }
    yield x;
}

function stepSplitStr(str, start, step) {
    var res = "";
    for (var i = start; i < str.length; i += step) {
        res = res.concat(str.charAt(i));
    }
    return res;
}

function encrypt(ptext, key) {
    const base = "A".charCodeAt(0);

    ptext = ptext.replace(/[^a-zA-Z]/g, "").toUpperCase();
    key = key.replace(/[^a-zA-Z]/g, "").toUpperCase();

    if (ptext === "" || key === "") {
        return "";
    }

    var ctext = "";

    for (let i = 0; i < ptext.length; i++) {
        var pi = ptext.charCodeAt(i) - base;
        var ki = key.charCodeAt(i % key.length) - base;
        var ci = ((pi + ki) % 26) + base;
        ctext = ctext.concat(String.fromCharCode(ci));
    }

    return ctext;
}

function decrypt(ctext, key) {
    const base = "A".charCodeAt(0);

    ctext = ctext.replace(/[^a-zA-Z]/g, "").toUpperCase();
    key = key.replace(/[^a-zA-Z]/g, "").toUpperCase();

    if (ctext === "" || key === "") {
        return "";
    }

    var ptext = "";

    for (let i = 0; i < ctext.length; i++) {
        var ci = ctext.charCodeAt(i) - base;
        var ki = key.charCodeAt(i % key.length) - base;
        var pi = ((ci - ki + 26) % 26) + base;
        ptext = ptext.concat(String.fromCharCode(pi));
    }

    return ptext;
}

function kasiskiExamination(ctext, sequence_len = 3) {
    ctext = ctext.replace(/[^a-zA-Z]/g, "").toUpperCase();

    if (ctext === "") {
        return {};
    }

    var sequences_dict = {};
    var factor_dict = {};

    for (let index = 0; index < ctext.length - sequence_len + 1; index++) {
        var current_sequence = ctext.slice(index, index + sequence_len);
        if (sequences_dict.hasOwnProperty(current_sequence)) {
            sequences_dict[current_sequence].push(index);
        } else {
            sequences_dict[current_sequence] = [index];
        }
    }

    for (const seq in sequences_dict) {
        if (sequences_dict[seq].length === 1) {
            delete sequences_dict[seq];
        } else {
            for (const comb of k_combinations(sequences_dict[seq], 2)) {
                var current_spacing = Math.abs(comb[0] - comb[1]);
                for (const f of get_factors(current_spacing)) {
                    var factor_count = factor_dict[f] || 0;
                    factor_dict[f] = factor_count + 1;
                }
            }
        }
    }

    return factor_dict;
}

// http://en.wikipedia.org/wiki/Letter_frequency
const ENGLISH_LETTER_FREQ = {
    'E': 12.70,
    'T': 9.06,
    'A': 8.17,
    'O': 7.51,
    'I': 6.97,
    'N': 6.75,
    'S': 6.33,
    'H': 6.09,
    'R': 5.99,
    'D': 4.25,
    'L': 4.03,
    'C': 2.78,
    'U': 2.76,
    'M': 2.41,
    'W': 2.36,
    'F': 2.23,
    'G': 2.02,
    'Y': 1.97,
    'P': 1.93,
    'B': 1.29,
    'V': 0.98,
    'K': 0.77,
    'J': 0.15,
    'X': 0.15,
    'Q': 0.10,
    'Z': 0.07,
}

function frequency_analysis(ctext) {
    ctext = ctext.replace(/[^a-zA-Z]/g, "").toUpperCase();

    var char_freq = {};
    ctext.split('').forEach(val => char_freq[val] = (char_freq[val] || 0) + 1);

    var score_dict = {};
    const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
    for (var i = 0; i < chars.length; i++) {
        var key = chars.charAt(i);
        score_dict[key] = 0;
        for (const c in char_freq) {
            var char_diff = c.charCodeAt(0) - key.charCodeAt(0);
            if (char_diff >= 0) {
                var lettr = char_diff % 26;
            } else {
                var lettr = 26 - (Math.abs(char_diff) % 26);
            }
            Math.abs() % 26;

            score_dict[key] += (
                char_freq[c] * 100 / ctext.length * ENGLISH_LETTER_FREQ[String.fromCharCode(lettr + 'A'.charCodeAt(0))]
            );
        }
    }

    return score_dict;
}

function keyDetection(ctext, key_len) {
    ctext = ctext.replace(/[^a-zA-Z]/g, "").toUpperCase();

    key_chars_candidate = []
    for (var i = 0; i < key_len; i++) {
        var sub_ctext = stepSplitStr(ctext, i, key_len);
        var score_dict = frequency_analysis(sub_ctext);
        var sorted_scores = [];
        for (var z in score_dict) {
            sorted_scores.push([z, score_dict[z]]);
        }
        sorted_scores.sort(function (a, b) {
            return b[1] - a[1];
        });
        var max_score = sorted_scores[0][1];
        var candids = [];
        for (var j = 0; j < sorted_scores.length; j++) {
            if (sorted_scores[j][1] >= (max_score - (max_score * 20 / 100))) {
                candids.push(sorted_scores[j][0]);
            }
            else {
                break;
            }
        }
        key_chars_candidate.push(candids);
    }

    return key_chars_candidate;
}

File: vigenere.py
import re
from collections import Counter
from itertools import combinations, product


def __get_factors(x: int):
    for i in range(2, (x // 2) + 1):
        if x % i == 0:
            yield i
    yield x


def __normalize(text):
    return re.sub("[^a-zA-Z]", "", text).upper()


def encrypt(ptext: str, key: str) -> str:
    base = ord("A")

    ptext = __normalize(ptext)
    key = __normalize(key)

    if ptext == "" or key == "":
        return ""

    ctext = ""

    for i in range(len(ptext)):
        pi = ord(ptext[i]) - base
        ki = ord(key[i % len(key)]) - base
        ci = ((pi + ki) % 26) + base
        ctext += chr(ci)

    return ctext


def decrypt(ctext: str, key: str) -> str:
    base = ord("A")

    ctext = __normalize(ctext)
    key = __normalize(key)

    if ctext == "" or key == "":
        return ""

    ptext = ""

    for i in range(len(ctext)):
        ci = ord(ctext[i]) - base
        ki = ord(key[i % len(key)]) - base
        pi = ((ci - ki + 26) % 26) + base
        ptext += chr(pi)

    return ptext


# http://en.wikipedia.org/wiki/Letter_frequency
ENGLISH_LETTER_FREQ = {
    'E': 12.70,
    'T': 9.06,
    'A': 8.17,
    'O': 7.51,
    'I': 6.97,
    'N': 6.75,
    'S': 6.33,
    'H': 6.09,
    'R': 5.99,
    'D': 4.25,
    'L': 4.03,
    'C': 2.78,
    'U': 2.76,
    'M': 2.41,
    'W': 2.36,
    'F': 2.23,
    'G': 2.02,
    'Y': 1.97,
    'P': 1.93,
    'B': 1.29,
    'V': 0.98,
    'K': 0.77,
    'J': 0.15,
    'X': 0.15,
    'Q': 0.10,
    'Z': 0.07,
}


def kasiski_examination(ctext, sequence_len=3):
    """
    Kasiski Examination: https://en.wikipedia.org/wiki/Kasiski_examination
    first, find every repeated set of letters at least three letters long
    in the ciphertext and then get Factors of Spacings and count them.
    """

    ctext = __normalize(ctext)

    # key is sequence and value is a list of all start index of sequence in
    # cipher text
    sequences_dict = {}

    # find all sub sequences with at least three letters
    for index in range(len(ctext) - sequence_len + 1):
        current_sequence = ctext[index : index + sequence_len]
        if sequences_dict.get(current_sequence):
            sequences_dict.get(current_sequence).append(index)
        else:
            sequences_dict[current_sequence] = [index]

    factor_dict = {}
    # calculate spacing between repeat sequences
    for seq in list(sequences_dict.keys()):
        if len(sequences_dict[seq]) == 1:
            # we don't need sequences without any repetition
            del sequences_dict[seq]
        else:
            for comb in combinations(sequences_dict[seq], 2):
                current_spacing = abs(comb[0] - comb[1])
                for f in __get_factors(current_spacing):
                    factor_dict[f] = factor_dict.get(f, 0) + 1
    return sorted(factor_dict.items(), key=lambda x: x[1], reverse=True)


def key_detection(ctext, key_len):
    ctext = __normalize(ctext)

    # get Every key_len'th letter starting with the i letter. they are
    # encoded with same char & it's Caesar cipher
    key_chars_candidate = []
    for i in range(key_len):
        score_dict = frequency_analysis(ctext[i::key_len])
        sorted_scores = sorted(score_dict.items(), key=lambda x: x[1], reverse=True)
        max_score = sorted_scores[0][1]
        candids = []
        for i, s in sorted_scores:
            if s >= (max_score - (max_score * 20 / 100)):
                candids.append(i)
            else:
                break
        key_chars_candidate.append(candids)

    return list(product(*key_chars_candidate))


def frequency_analysis(ctext):
    ctext = __normalize(ctext)

    char_freq = Counter(ctext)
    score_dict = {}
    for key in 'ABCDEFGHIJKLMNOPQRSTUVWXYZ':
        score_dict[key] = 0
        for c in char_freq:
            score_dict[key] += (
                char_freq[c] * 100 / len(ctext) * ENGLISH_LETTER_FREQ[chr((ord(c) - ord(key)) % 26 + ord('A'))]
            )

    return score_dict
